x/auth/middleware/recovery.go

```diff
@@ -54,19 +59,18 @@ func (txh recoveryTxHandler) DeliverTx(ctx context.Context, req tx.Request) (res
 		}
 	}()
 
-	// If BlockGasMeter() panics it will be caught by the above recover and will
-	// return an error - in any case BlockGasMeter will consume gas past the limit.
-	//
-	// NOTE: This must exist in a separate defer function for the above recovery
-	// to recover from this one.
+	// Must not panic in this defer function, because tx is already executed and committed,
+	// panic here will return error result for a successfully executed tx.
 	defer func() {
+		gasConsumed := sdkCtx.GasMeter().GasConsumedToLimit()
+		remaining := sdkCtx.BlockGasMeter().GasRemaining()
+		if gasConsumed > remaining {
+			gasConsumed = remaining
+		}
+		// there should be no danger of panic or overflow
 		sdkCtx.BlockGasMeter().ConsumeGas(
-			sdkCtx.GasMeter().GasConsumedToLimit(), "block gas meter",
+			gasConsumed, "block gas meter",
```

Comment:
> only consume remaining gases when block gas limit exceeded

Agreed for the fact we shouldn't panic here. But are we still including the tx? How do we not include it?

Reply:
Yes, the tx was and still is included. The "issue" was the fact that the tx was included in a block but this could've paniced which would lead the user to believe it wasn't.

Reply:
Ah understood, thanks 👍

Reply:
Actually, follow-up question after reading Ethan's msg: why are we are not returning an error here?

```diff
		if gasConsumed > remaining {
-			gasConsumed = remaining
+			return sdkerrors.ErrOutOfGas("block gas meter exhausted")
		}
```

?

Reply:
> How do we not include it?

We added a tx gas limit check before tx execution.

Reply:
> why are we are not returning an error here?

Because here the tx is already executed and committed, return error will not revert tx side-effects.

Maybe a cleaner solution is to use a `CacheContext` here?

Reply:
> why are we are not returning an error here?

at here, the tx msgs are already run and committed.

Reply:
> at here, the tx msgs are already run and committed.

Yes okay, you're right. Moving the cacheContext our from runMsgsHandler to a outer SnapshotMiddlewar makes sense to me